File: csharp/Tdcv2/Generators/Stat.cs
using System.Globalization;

namespace Tdcv2.Generators;

/// <summary>
/// <c>&lt;gen type="stat"&gt;</c> — one number for the WHOLE run, on every row.
/// </summary>
/// <remarks>
/// <para><c>accumulate=</c> totals a list inside one record. <c>&lt;gen type="running"&gt;</c>
/// totals a column as it goes, so row i knows about rows 1..i. This is the third and last axis:
/// a row that knows something about EVERY row, including the ones after it.</para>
///
/// <para><c>sum</c>, <c>min</c> and <c>max</c> are the last value of the corresponding RUNNING
/// column, computed by <see cref="Accumulate.ApplyColumn"/>. That is not a shortcut — it is how
/// the two features are kept from drifting: the fixed-point scale rule, the treatment of an empty
/// cell and the "min returns the winning element's own spelling" rule are written once and used
/// twice.</para>
///
/// <para><c>mean</c>, <c>median</c> and <c>stddev</c> are ratios and cannot be exact, so they are
/// computed in floating point over the numeric values — the same three formulas the expression
/// language's list functions use, including the POPULATION standard deviation. <c>decimals=</c>
/// rounds the answer through the same ToFixed <c>decimals=</c> on a number already uses.</para>
/// </remarks>
internal static class Stat
{
    /// <summary>What a statistic can be.</summary>
    internal static readonly string[] Ops =
        { "sum", "mean", "median", "min", "max", "count", "stddev" };

    /// <summary>
    /// Read <c>op=</c> where an unknown op simply means "none".
    /// </summary>
    /// <remarks>
    /// The engine path uses this one: by the time a value is drawn the validator has already
    /// refused a misspelled op, so throwing here would turn a reported problem into a crash.
    /// </remarks>
    internal static string? ReadOp(IReadOnlyDictionary<string, string> attrs)
    {
        string raw = (attrs.GetValueOrDefault("op") ?? "").Trim();
        return Array.IndexOf(Ops, raw) >= 0 ? raw : null;
    }

    /// <summary>The same, but strict — the validator's copy, which turns a bad op into a diagnostic.</summary>
    internal static string? ParseOp(IReadOnlyDictionary<string, string> attrs)
    {
        string raw = (attrs.GetValueOrDefault("op") ?? "").Trim();
        if (raw.Length == 0)
        {
            return null;
        }

        if (Array.IndexOf(Ops, raw) < 0)
        {
            throw new StatException($"op=\"{raw}\" is not one of {string.Join(", ", Ops)}");
        }

        return raw;
    }

    /// <summary><c>decimals=</c>, or null when the answer is printed at full precision.</summary>
    internal static int? ParseDecimals(IReadOnlyDictionary<string, string> attrs)
    {
        string raw = (attrs.GetValueOrDefault("decimals") ?? "").Trim();
        if (raw.Length == 0)
        {
            return null;
        }

        if (!int.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out int n)
            || n < 0 || n > 10)
        {
            throw new StatException($"decimals=\"{raw}\" is not a whole number from 0 to 10");
        }

        return n;
    }

    /// <summary>
    /// The statistic itself, as the text that goes in every cell.
    /// </summary>
    /// <remarks>
    /// A cell the parent filter emptied does not take part — the same rule
    /// <see cref="Accumulate.ApplyColumn"/> follows, so a filtered column has one meaning across
    /// the three features rather than three.
    /// </remarks>
    internal static string Statistic(string?[] values, string op, int? decimals)
    {
        List<string> present = values
            .Where(v => v is not null && v.Trim().Length != 0)
            .Select(v => v!)
            .ToList();
        if (op == "count")
        {
            return present.Count.ToString(CultureInfo.InvariantCulture);
        }

        if (present.Count == 0)
        {
            return string.Empty;
        }

        if (op is "sum" or "min" or "max")
        {
            // The last value of the running column IS the total over every row, and reusing it
            // is what keeps the exact-decimal arithmetic from drifting.
            string?[] running = Accumulate.ApplyColumn(values, op, null, null);
            string last = running.LastOrDefault(v => v is not null) ?? string.Empty;
            return decimals is null ? last : Fixed(AsNumber(last), decimals.Value);
        }

        double[] figures = present.Select(AsNumber).ToArray();
        double answer = op switch
        {
            "mean" => Mean(figures),
            "median" => Median(figures),
            _ => StdDev(figures),
        };
        return decimals is null ? ToText(answer) : Fixed(answer, decimals.Value);
    }

    /// <summary>
    /// A double as JavaScript prints it — a whole number without a decimal point.
    /// </summary>
    /// <remarks>
    /// The reference writes <c>String(x)</c> and the four ports each imitate it; this is C#'s
    /// copy, kept beside its one caller rather than added to a shared helper nothing else needs.
    /// </remarks>
    private static string ToText(double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value) || value != Math.Floor(value))
        {
            return value.ToString("R", CultureInfo.InvariantCulture);
        }

        return ((long)value).ToString(CultureInfo.InvariantCulture);
    }

    /// <summary>A cell as a number. The column it reads is numeric by construction.</summary>
    private static double AsNumber(string raw) =>
        double.TryParse(raw.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double v)
            ? v
            : double.NaN;

    private static double Mean(double[] values) => values.Sum() / values.Length;

    private static double Median(double[] values)
    {
        double[] sorted = (double[])values.Clone();
        Array.Sort(sorted);
        int half = sorted.Length / 2;
        return sorted.Length % 2 == 1 ? sorted[half] : (sorted[half - 1] + sorted[half]) / 2;
    }

    /// <summary>
    /// The POPULATION standard deviation — divided by n, matching <c>stddev()</c> in an expression.
    /// </summary>
    private static double StdDev(double[] values)
    {
        double average = Mean(values);
        double variance = values.Sum(v => (v - average) * (v - average)) / values.Length;
        return Maths.TdcMath.Sqrt(variance);
    }

    /// <summary>
    /// <c>decimals=</c> applied.
    /// </summary>
    /// <remarks>
    /// The same ToFixed <c>decimals=</c> on <c>&lt;gen type="number"&gt;</c> already uses, and
    /// nothing hand-rolled: multiplying by 10^decimals and flooring introduces a rounding error of
    /// its own before the rounding rule ever runs, so two implementations could land on either
    /// side of a tie for the same input.
    /// </remarks>
    private static string Fixed(double value, int decimals) =>
        double.IsNaN(value) || double.IsInfinity(value)
            ? ToText(value)
            : Stats.Distribution.ToFixed(value, decimals);
}

/// <summary>A statistic that cannot be read as one.</summary>
internal sealed class StatException : Exception
{
    internal StatException(string message)
        : base(message)
    {
    }
}

File: python/src/tdcv2/generators/stat.py
"""``<gen type="stat">`` — one number for the WHOLE run, on every row.

``accumulate=`` totals a list inside one record. ``<gen type="running">`` totals a column as it
goes, so row i knows about rows 1..i. This is the third and last axis: a row that knows something
about EVERY row, including the ones after it.

``sum``, ``min`` and ``max`` are the last value of the corresponding RUNNING column, computed by
:func:`tdcv2.generators.accumulate.apply_column`. That is not a shortcut — it is how the two
features are kept from drifting: the fixed-point scale rule, the treatment of an empty cell and
the "min returns the winning element's own spelling" rule are written once and used twice.

``mean``, ``median`` and ``stddev`` are ratios and cannot be exact, so they are computed in
floating point over the numeric values — the same three formulas the expression language's list
functions use, including the POPULATION standard deviation. ``decimals=`` rounds the answer;
without it the full value is printed, because a mean that quietly lost digits is worse than an
ugly one.
"""

from __future__ import annotations

import math

from ..lib import numbers
from ..math import tdc_math
from . import accumulate as accumulate_gen

OPS = ("sum", "mean", "median", "min", "max", "count", "stddev")


class StatError(ValueError):
    """A statistic that cannot be read as one."""


def read_op(attrs: dict[str, str]) -> str | None:
    """Read ``op=`` where an unknown op simply means "none".

    The engine path uses this one: by the time a value is drawn the validator has already refused
    a misspelled op, so throwing here would turn a reported problem into a crash.
    """
    raw = (attrs.get("op") or "").strip()
    return raw if raw in OPS else None


def parse_op(attrs: dict[str, str]) -> str | None:
    """The same, but strict — the validator's copy, which turns a bad op into a diagnostic."""
    raw = (attrs.get("op") or "").strip()
    if raw == "":
        return None
    if raw not in OPS:
        raise StatError(f'op="{raw}" is not one of {", ".join(OPS)}')
    return raw


def parse_decimals(attrs: dict[str, str]) -> int | None:
    """``decimals=``, or None when the answer is printed at full precision."""
    raw = (attrs.get("decimals") or "").strip()
    if raw == "":
        return None
    try:
        n = int(raw)
    except ValueError:
        raise StatError(f'decimals="{raw}" is not a whole number from 0 to 10') from None
    if str(n) != raw.lstrip("+") or n < 0 or n > 10:
        raise StatError(f'decimals="{raw}" is not a whole number from 0 to 10')
    return n


def statistic(values: list[str | None], op: str, decimals: int | None) -> str:
    """The statistic itself, as the text that goes in every cell.

    A cell the parent filter emptied does not take part — the same rule ``apply_column`` follows,
    so a filtered column has one meaning across the three features rather than three.
    """
    present = [v for v in values if v is not None and v.strip() != ""]
    if op == "count":
        return str(len(present))
    if not present:
        return ""

    if op in ("sum", "min", "max"):
        # The last value of the running column IS the total over every row, and reusing it is
        # what keeps the exact-decimal arithmetic from drifting.
        running = accumulate_gen.apply_column(values, op, None, None)
        last = next((v for v in reversed(running) if v is not None), "")
        return last if decimals is None else _fixed(float(last), decimals)

    figures = [float(v) for v in present]
    if op == "mean":
        answer = _mean(figures)
    elif op == "median":
        answer = _median(figures)
    else:
        answer = _stddev(figures)
    return numbers.to_text(answer) if decimals is None else _fixed(answer, decimals)


def _mean(values: list[float]) -> float:
    return sum(values) / len(values)


def _median(values: list[float]) -> float:
    ordered = sorted(values)
    half = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[half]
    return (ordered[half - 1] + ordered[half]) / 2


def _stddev(values: list[float]) -> float:
    """The POPULATION standard deviation — divided by n, matching ``stddev()`` in an expression."""
    average = _mean(values)
    variance = sum((v - average) * (v - average) for v in values) / len(values)
    return tdc_math.sqrt(variance)


def _fixed(value: float, decimals: int) -> str:
    """``decimals=`` applied.

    ``numbers.to_fixed`` and nothing hand-rolled, deliberately. Multiplying by 10^decimals and
    flooring introduces a rounding error of its own before the rounding rule ever runs, so two
    implementations could land on either side of a tie for the same input. ``to_fixed`` works on
    the decimal expansion of the double itself and is what ``decimals=`` on ``<gen type="number">``
    already uses, so the attribute means one thing across the whole engine rather than two.
    """
    if math.isnan(value) or math.isinf(value):
        return numbers.to_text(value)
    return numbers.to_fixed(value, decimals)
